src/localization.rs

```diff
@@ -0,0 +1,361 @@
+    ///
+    /// In general, this should not be used directly; [`LocalizedString`]
+    /// should be used for localization, and you should call
+    /// [`LocalizedString::resolve`] to update the string as required.
+    ///
+    ///[`LocalizedString`]: struct.LocalizedString.html
+    ///[`LocalizedString::resolve`]: struct.LocalizedString.html#method.resolve
+    pub fn localize<'args>(
+        &'args self,
+        key: &str,
+        args: impl Into<Option<&'args FluentArgs<'args>>>,
+    ) -> Option<String> {
+        let args = args.into();
+        let value = match self
+            .current_bundle
+            .get_message(key)
+            .and_then(|msg| msg.value)
+        {
+            Some(v) => v,
+            None => return None,
+        };
+        let mut errs = Vec::new();
+        let result = self.current_bundle.format_pattern(value, args, &mut errs);
+        for err in errs {
+            eprintln!("localization error {:?}", err);
+        }
+        Some(result.to_string())
+    }
+    //TODO: handle locale change
+}
+
+impl<T> LocalizedString<T> {
+    /// Create a new `LocalizedString` with the given key.
+    pub const fn new(key: &'static str) -> Self {
+        LocalizedString {
+            key,
+            args: None,
+            placeholder: None,
+            resolved: None,
+            resolved_lang: None,
+        }
+    }
+
+    /// Add a placeholder value. This will be used if localization fails.
+    ///
+    /// This is intended for use during prototyping.
+    pub const fn with_placeholder(mut self, placeholder: &'static str) -> Self {
+        self.placeholder = Some(placeholder);
+        self
+    }
+
+    /// Return the localized value for this string, or the placeholder, if
+    /// the localization is missing, or the key if there is no placeholder.
+    fn current_value(&self) -> &str {
+        self.resolved
+            .as_ref()
+            .map(|s| s.as_str())
+            .or(self.placeholder)
+            .unwrap_or(self.key)
+    }
+}
+
+impl<T: Data> LocalizedString<T> {
+    /// Add a named argument and a cooresponding [`ArgClosure`]. This closure
+    /// is a function that will return a value for the given key from the current
+    /// environment and data.
+    ///
+    /// [`ArgClosure`]: type.ArgClosure.html
+    pub fn with_arg(
+        mut self,
+        key: &'static str,
+        f: impl Fn(&Env, &T) -> FluentValue<'static> + 'static,
+    ) -> Self {
+        self.args
+            .get_or_insert(HashMap::new())
+            .insert(key, ArgSource(Arc::new(f)));
+        self
+    }
+
+    /// Lazily compute the localized value for this string based on the provided
+    /// environment and data.
+    pub fn resolve<'a>(&'a mut self, env: &Env, data: &T) -> &'a str {
+        //TODO: this recomputes the string if either the language has changed,
+        //or *anytime* we have arguments. Ideally we would be using a lens
```

Comment:
I don't think we have to do anything with lenses here, this logic can piggy back on just the changes to `T`. The way to do that is to have `paint` call a method that resolves the string iff the resolution hasn't been done yet, otherwise just fetches the resolution, then have `update` recompute the string. As I said below, even better is if the "recompute" method lets us decide whether to propagate an invalidation. Just redoing the string every time `T` changes seems cheap enough to me.

Reply:
My thinking here is that with something like lenses, you could indicate per-arg where in the data tree it comes from, and then we could stash that and do better invalidation automatically by comparing whether those items have changed. I'm not sure how much of a saving this this, but would definitely be a handful fewer allocations per localized string, per update.